src/CloudNative.CloudEvents.SystemTextJson/JsonEventFormatter.cs

```diff
@@ -120,31 +120,72 @@ public JsonEventFormatter(JsonSerializerOptions serializerOptions, JsonDocumentO
+            DecodeBatchModeMessageImpl(new MemoryStream(data), contentType, extensionAttributes, false).GetAwaiter().GetResult();
+
+        private async Task<IReadOnlyList<CloudEvent>> DecodeBatchModeMessageImpl(Stream data, ContentType contentType, IEnumerable<CloudEventAttribute> extensionAttributes, bool async)
+        {
+            data = data ?? throw new ArgumentNullException(nameof(data));
+            var document = await ReadDocumentAsync(data, contentType, async).ConfigureAwait(false);
+            using (document)
+            {
+                var root = document.RootElement;
+                if (root.ValueKind != JsonValueKind.Array)
+                {
+                    throw new ArgumentException($"Cannot decode JSON element of kind '{root.ValueKind}' as batch CloudEvent");
+                }
+                // Avoiding LINQ to avoid extraneous allocations etc.
+                List<CloudEvent> events = new List<CloudEvent>(root.GetArrayLength());
+                foreach (var element in root.EnumerateArray())
+                {
+                    events.Add(DecodeJsonElement(element, extensionAttributes));
+                }
+                return events;
+            }
+        }
+
+        private async Task<JsonDocument> ReadDocumentAsync(Stream data, ContentType contentType, bool async)
+        {
             var encoding = contentType.GetEncoding();
-            JsonDocument document;
             if (encoding is UTF8Encoding)
             {
-                document = async
+                return async
                     ? await JsonDocument.ParseAsync(data, DocumentOptions).ConfigureAwait(false)
                     : JsonDocument.Parse(data, DocumentOptions);
             }
             else
             {
                 using var reader = new StreamReader(data, encoding);
-                string json = async
+                var json = async
                     ? await reader.ReadToEndAsync().ConfigureAwait(false)
                     : reader.ReadToEnd();
-                document = JsonDocument.Parse(json, DocumentOptions);
-            }
-            using (document)
-            {
-                return DecodeJsonDocument(document, extensionAttributes);
+                return JsonDocument.Parse(json, DocumentOptions);
             }
         }
 
-        private CloudEvent DecodeJsonDocument(JsonDocument document, IEnumerable<CloudEventAttribute> extensionAttributes = null)
+        // TODO: Override the other methods
+
+        private CloudEvent DecodeJsonElement(JsonElement element, IEnumerable<CloudEventAttribute> extensionAttributes = null)
         {
-            if (!document.RootElement.TryGetProperty(CloudEventsSpecVersion.SpecVersionAttribute.Name, out var specVersionProperty)
+            if (element.ValueKind != JsonValueKind.Object)
```

Comment:
What about a single element array?

Reply:
I personally think trying to parse a single-element batch as if it's not a batch at all should still fail. The calling code is doing the wrong thing, and that should be reported as early as possible - otherwise I can see code that works in all tests and then fails with real batches. Happy to be persuaded though.

Reply:
I was thinking of the case where a caller might know that they will always get a single event in a JSON array. For instance, maybe the publisher always publishes it this way. But I think you raise a fair point about the potential for missing this during testing.

Reply:
Yup - will leave it for now. I suspect we may want more introspection capabilities on event formatters before we go GA, so you could have multiple formatters and pick the right one to use (and the right method to use, potentially) based on the content type.